lib/puppet/pops/validation/checker4_0.rb

```diff
@@ -530,6 +533,80 @@ def internal_check_future_reserved_word(o, name)
     end
   end
 
+  def internal_check_file_namespace(o, name, file)
+    return if file.nil? || file == '' #e.g. puppet apply -e '...'
+
+    lc_file = file.downcase
+
+    file_namespace = @file_to_namespace[lc_file]
+    if file_namespace.nil?
+      return if @file_to_namespace.key?(lc_file)
+      file_namespace = @file_to_namespace[lc_file] = namespace_for_file(lc_file)
+      return if file_namespace.nil?
+    end
+
+    if !name.downcase.start_with?(file_namespace)
+      acceptor.accept(Issues::ILLEGAL_DEFINITION_LOCATION, o, {:name => name, :file => file})
+    end
+  end
+
+  NEVER_MATCH = '\b'.freeze
+
+  def namespace_for_file(file)
+    path = Pathname.new(file)
+
+    return nil if path.extname != ".pp"
+
+    return nil if initial_manifest?(path)
+
+    path = path.each_filename.to_a
+
+    # Example definition dir: manifests in this path:
+    # <modules dir>/<module name>/manifests/<module subdir>/<classfile>.pp
+    definition_dir_index = find_module_definition_dir(path)
+
+    # How can we get this result?
+    # If it is not an initial manifest, it must come from a module,
+    # and from the manifests dir there.  This may never get used...
+    return NEVER_MATCH if definition_dir_index.nil? || definition_dir_index == 0
+
+    before_definition_dir = definition_dir_index - 1
+    after_definition_dir = definition_dir_index + 1
+    names = path[after_definition_dir .. -2] # Directories inside module
+    names.unshift(path[before_definition_dir]) # Name of the module itself
+    # Do not include name of module init file at top level of module
+    filename = path[-1]
+    if !(path.length == (after_definition_dir+1) && filename == 'init.pp')
+      names.push(filename[0 .. -4]) # Remove .pp from filename
+    end
+
+    names.join("::").freeze
+  end
+
+  def initial_manifest?(path)
+    env = Puppet.lookup(:current_environment)
+    manifest_setting = env.nil? ? nil : env.manifest
+
+    # Does this ever happen outside of tests?
+    if manifest_setting.nil? || manifest_setting == :no_manifest
+      return path.basename.to_s == 'site.pp'
```

Comment:
This looks strange - which case is this? It seems it will allow any `site.pp` file (in any location) to contain any name. A `site.pp` is really only special in that it is the default for manifest in an environment.